CHC solver access

maul-esel opened the pull request:
This PR allows access to several CHC solvers via a unified API. Solvers can be invoked via the `IChcScript` interface, or by including the `ChcSolver` plugin in the toolchain (the plugin just calls `IChcScript` methods internally). We have some support for extracting models (if SAT) and derivations or unsat cores (if UNSAT).

-  **[eldarica](https://github.com/uuverifiers/eldarica)**, accessed via the scala API. Supports models, derivations and UNSAT cores.
-  **[golem](https://github.com/usi-verification-and-security/golem)**, executed via command line. Supports models.
  *Note:* Currently, there is no Windows build of golem. I've spent some time looking into it, and it seems like it might be achievable, but no guarantees.
- **[Z3 / spacer](https://github.com/Z3Prover/z3)**, via the SMT-LIB2 interface (`Script`). Supports models and UNSAT cores.
- **TreeAutomizer**, by starting the CEGAR loop directly. Supports UNSAT cores.

## TODO
- [x] reorganize the regression tests: drop tests that time out or use features unsupported by the solvers (we don't intend to test the solvers here, only their integration)

maul-esel committed:
a first attempt at using eldarica API

maul-esel committed:
fix classpath issues

maul-esel committed:
fixes & extensions for expression translation

maul-esel committed:
eldarica: derivation is available as DAG

maul-esel committed:
add class for derivations

maul-esel committed:
extract eldarica derivations in case of UNSAT

maul-esel committed:
reorganize eldarica API access

Split translation and backtranslation into two separate classes.

maul-esel committed:
sketch an interface for CHC solvers

maul-esel committed:
rename variable for consistency

maul-esel committed:
begin implementing IChcScript for eldarica

maul-esel committed:
begin IChcScript implementation based on Script (for z3)

maul-esel committed:
sketch IChcScript implementation for TreeAutomizer

maul-esel committed:
add TODO

maul-esel committed:
preliminary work on UNSAT cores from TreeAutomizer

maul-esel committed:
NamedTermWrapper: fix method name

maul-esel committed:
add HcSymbolTable to IChcScript::solve

maul-esel committed:
fix comment

maul-esel committed:
SmtChcScript: declare predicate symbols

maul-esel committed:
Merge branch 'dev' into wip/dk/eldarica-scala

maul-esel committed:
add ChcSolver plugin

maul-esel committed:
import TreeAutomizer to get access to TreeAutomizerChcScript

maul-esel committed:
eldarica access: remove SimpleAPI from public API

maul-esel committed:
begin sketching ChcSolver plugin

maul-esel committed:
extended docs on IChcScript, make possibility of absent model / derivation / UNSAT core explicit

maul-esel committed:
avoid code duplication for SMT printing of Horn clauses

maul-esel committed:
add timeout to IChcScript API (not yet implemented)

maul-esel committed:
simplify SmtChcScript

maul-esel committed:
begin sketching golem CHC script

maul-esel committed:
add preferences for ChcSolver plugin

maul-esel committed:
rename Eldarica class

maul-esel committed:
BidirectionalMap: add support for computeIfAbsent

maul-esel committed:
more backtranslation work for eldarica

maul-esel committed:
rename Horn clause SAT/UNSAT results for use by all solvers

maul-esel committed:
move CHC results to Library-CHC

maul-esel committed:
return empty Optional, not null

maul-esel committed:
ChcSolver: report results

maul-esel committed:
add a test suite for ChcSolver

maul-esel committed:
add timeout support for EldaricaChcScript

Allow setting a default timeout for queries, in the constructor.
When solve() is called with a timeout for the query, the default
timeout is ignored.

Additionally, we must take the toolchain timeout into account.

maul-esel committed:
some output cleanup in TreeAutomizerTestResultDecider

maul-esel committed:
eldarica: more work on translation and backtranslation

maul-esel committed:
TreeAutomizerOverallResultEvaluator: distinguish CRASH from UNKNOWN

maul-esel committed:
add ChcSolver to features

maul-esel committed:
prepare settings files for ChcSolver

maul-esel committed:
ChcSolverObserved: provide ManagedScript to SmtChcScript

maul-esel committed:
SmtChcScript: several fixes

- assume the caller has already set the script's logic
- unlock the script for ChcAsserter
- fix handling of UNSAT core term names

maul-esel committed:
SmtChcScript: assume function symbols already declared in script

maul-esel committed:
add a more suitable IResult for UNKNOWN

UnprovableResult is supposed to be specific to a location,
and have an execution. Neither is applicable here.

maul-esel committed:
HornClause: make sure to actually name terms if requested

maul-esel committed:
prepare timeout support for golem

maul-esel committed:
Merge branch 'dev' into wip/dk/chc-solvers

# Conflicts:
#	trunk/source/BA_FeatureUltimateCommon/feature.xml

maul-esel committed:
add chcsolver to launch config

maul-esel committed:
rename regression test suite and add settings for ChcSolver regression tests

maul-esel committed:
delete obsolete test suite

danieldietsch committed:
add dependency to chcsolver plugin

maul-esel committed:
let ChcSolver plugin return the computed solution

maul-esel committed:
ChcUnsatResult: derivation and/or unsat core instead of Object

maul-esel committed:
cleanup of Chc(Un)SatResult, allow adding model to result

maul-esel committed:
extend script result grammar: golem does not put parentheses around model

maul-esel committed:
implement golem support via Executor

maul-esel committed:
properly handle eldarica timeouts

maul-esel committed:
TreeAutomizer: add message to assert

maul-esel committed:
add current golem version (0.4.1)

schuessf committed:
Add some thread modular tests

schuessf committed:
Fix verdict and rename file

maul-esel committed:
eldarica translation: add support for boolean literals

maul-esel committed:
add some timeout handling to TreeAutomizer

maul-esel committed:
eldarica: also support boolean literals in terms; remove unused code

maul-esel committed:
fix eldarica tiemout logic: avoid negative timeouts

maul-esel committed:
eldarica: remember last result if it was UNKNOWN

maul-esel committed:
EldaricaChcScript: add some logging

maul-esel committed:
simplify FunctionDefinition: return sort is part of function symbol

maul-esel committed:
fix FunctionDefinition::toString

maul-esel committed:
EldaricaChcScript: properly implement getModel()

maul-esel committed:
add eldarica license

maul-esel committed:
use SMTLIBConstants

maul-esel committed:
better exception message

maul-esel committed:
remove obsolete code

maul-esel committed:
implement timeout support for TreeAutomizerChcScript

maul-esel committed:
SmtChcScript: implement solve with timeout

maul-esel committed:
eldarica Backtranslator: finish bool handling

maul-esel committed:
golem: dynamically determine script path

maul-esel committed:
adjust comment

schuessf reviewed (approved):
Looks good to me! Regarding the tests: If we just want to test the integration and not the solvers (which is not quite true for TreeAutomizer), we could just set `unknownIsSuccess` to `true` in `ChcRegressionTestSuite`.

schuessf commented on trunk/source/UltimateRegressionTest/src/de/uni_freiburg/informatik/ultimate/regressiontest/generic/ChcRegressionTestSuite.java:
Just a minor note: I guess we should also rename `TreeAutomizerTestResultDecider` and all classes that are used inside, since they are not exclusively used for Tree Automizer.

maul-esel committed:
rename test utility classes (suggested by @schuessf)

maul-esel commented:
Setting `unknownIsSuccess` to `true` is indeed problematic, because we still want to test TreeAutomizer. Also, for unsupported features, exceptions are thrown -- which shouldn't count as UNKNOWN.

danieldietsch commented on trunk/source/ChcSolver/src/de/uni_freiburg/informatik/ultimate/plugins/chcsolver/ChcSolverObserver.java:
```diff
@@ -0,0 +1,169 @@
+			mSolution = ChcSolution.unknown();
+			return new ChcUnknownResult(Activator.PLUGIN_ID, "CHC solver returned UNKNOWN.");
+		default:
+			throw new IllegalStateException();
```
Better UnsupportedOperationException with the unsupported value (or AssertionError)

danieldietsch commented on trunk/source/Library-CHC/src/de/uni_freiburg/informatik/ultimate/lib/chc/GolemChcScript.java:
```diff
@@ -0,0 +1,186 @@
+
+			dumperScript.exit();
+		} catch (final IOException e) {
+			throw new IllegalStateException(e);
```
why IllegalStateException? This has nothing to do with state, right? Just wrap it in a RuntimeException.

danieldietsch commented on trunk/source/Library-CHC/src/de/uni_freiburg/informatik/ultimate/lib/chc/GolemChcScript.java:
```diff
@@ -0,0 +1,186 @@
+			return mLastResult;
+
+		} catch (final IOException e) {
+			throw new IllegalStateException(e);
```
why IllegalStateException? This has nothing to do with state, right? Just wrap it in a RuntimeException.

danieldietsch commented on trunk/source/Library-CHC/src/de/uni_freiburg/informatik/ultimate/lib/chc/GolemChcScript.java:
```diff
@@ -0,0 +1,186 @@
+					null, null, null, determineTimeout(timeout));
+
+			mLastResult = executor.parseCheckSatResult();
+			switch (mLastResult) {
```
This switch looks out of place.
Why not
``mLastModel = (mProduceModels && mLastResult == SAT) ? executor.parseGetModelResult() : null;``

maul-esel replied:
The switch is probably leftover from when I considered UNSAT separately -- for parsing derivations or UNSAT cores. But as long as they are not supported, the ternary seems better, I agree.

danieldietsch commented on trunk/source/Library-CHC/src/de/uni_freiburg/informatik/ultimate/lib/chc/HornClause.java:
```diff
@@ -329,6 +323,20 @@ public Term constructFormula(final ManagedScript mgdScript, final boolean nameAs
 	}
 
+	private Term nameTerm(final ManagedScript mgdScript, final Term term) {
+		// TODO naming is a bit hacky
```
😅

danieldietsch commented on trunk/source/Library-CHC/src/de/uni_freiburg/informatik/ultimate/lib/chc/IChcScript.java:
```diff
@@ -0,0 +1,178 @@
+	 *
+	 * No other commands should be invoked on this script.
+	 *
+	 * @return
```
remove empty javadoc tag

danieldietsch commented on trunk/source/Library-CHC/src/de/uni_freiburg/informatik/ultimate/lib/chc/IChcScript.java:
```diff
@@ -0,0 +1,178 @@
+	/**
+	 * Try to solve the given system of Horn clauses.
+	 *
+	 * The instance may have an internal timeout, after which {@code UNKNOWN} is returned.
```
What is this internal timeout?

maul-esel replied:
E.g. when the solver is backed by Z3, the timeout set on the Z3 command line. Maybe that is too specific, and we should just note that the call may return `UNKNOWN`.

danieldietsch commented on trunk/source/Library-CHC/src/de/uni_freiburg/informatik/ultimate/lib/chc/IChcScript.java:
```diff
@@ -0,0 +1,178 @@
+	 *            A symbol table describing the predicates and other symbols used by the given CHC system
+	 * @param system
+	 *            The list of Horn clauses forming the CHC system
+	 * @param timeout
```
what is the valid range of the timeout? what happens with timeout=0? what happens if the timeout is reached?

maul-esel replied:
```
* @param timeout
*            The maximum time to be used for solving the system, in milliseconds. The timeout must be greater than
*            zero. If the timeout is reached before a solution is found, {@code UNKNOWN} is returned.
```

danieldietsch commented on trunk/source/Library-CHC/src/de/uni_freiburg/informatik/ultimate/lib/chc/IChcScript.java:
```diff
@@ -0,0 +1,178 @@
+	LBool solve(HcSymbolTable symbolTable, List<HornClause> system, long timeout);
+
+	/**
+	 * Determines if the solver class implements retrieval of models for {@code SAT} queries.
```
you mean calls to #solve that return SAT?

danieldietsch commented on trunk/source/Library-CHC/src/de/uni_freiburg/informatik/ultimate/lib/chc/IChcScript.java:
```diff
@@ -0,0 +1,178 @@
+	 * Model production must have been enabled via {@link #produceModels(boolean)} prior to the call to
+	 * {@link #solve(HcSymbolTable, List)}.
+	 *
+	 * @return
```
remove or describe

danieldietsch commented on trunk/source/Library-CHC/src/de/uni_freiburg/informatik/ultimate/lib/chc/IChcScript.java:
```diff
@@ -0,0 +1,178 @@
+	 * Derivation production must have been enabled via {@link #produceDerivations(boolean)} prior to the call to
+	 * {@link #solve(HcSymbolTable, List)}.
+	 *
+	 * @return
```
describe or remove

danieldietsch commented on trunk/source/Library-CHC/src/de/uni_freiburg/informatik/ultimate/lib/chc/IChcScript.java:
```diff
@@ -0,0 +1,178 @@
+	 * UNSAT core production must have been enabled via {@link #produceUnsatCores(boolean)} prior to the call to
+	 * {@link #solve(HcSymbolTable, List)}.
+	 *
+	 * @return
```
describe or remove

danieldietsch commented on trunk/source/Library-CHC/src/de/uni_freiburg/informatik/ultimate/lib/chc/SmtChcScript.java:
```diff
@@ -0,0 +1,157 @@
+		} finally {
+			try {
+				// reset timeout
+				mMgdScript.getScript().setOption(":timeout", 0L);
```
this looks _really_ hacky

maul-esel replied:
Yes. I don't know if there is a better solution, do you have an idea?

danieldietsch commented on trunk/source/Library-CHC/src/de/uni_freiburg/informatik/ultimate/lib/chc/SmtChcScript.java:
```diff
@@ -0,0 +1,157 @@
+
+	@Override
+	public Optional<Set<HornClause>> getUnsatCore() {
+		final var core = mMgdScript.getUnsatCore(this);
```
dont you want to check mProduceUnsatCores?

maul-esel replied:
I'm unsure. You mean we should throw an exception if it's `false` ?

Technically the behaviour of this method is unspecified if UNSAT cores were not previously enabled. It might be a good idea to catch such errors early though... But then I should probably check that consistently, i.e., also for `getModel()`, and similarly in `GolemChcScript`.

danieldietsch commented on trunk/source/Library-CHC/src/de/uni_freiburg/informatik/ultimate/lib/chc/eldarica/EldaricaChcScript.java:
```diff
@@ -0,0 +1,370 @@
+		mServices = services;
+		mLogger = services.getLoggingService().getLogger(getClass());
+		mScript = script;
+		mPrincess = SimpleAPI.apply(SimpleAPI.apply$default$1(), SimpleAPI.apply$default$2(),
```
what is happening here?

maul-esel replied:
An instance of the `SimpleAPI`-class is created, with default parameters (unfortunately, scala's default parameter values are somewhat verbose to use in Java). The class belongs to the API of eldarica's underlying SMT solver princess, which is used to create terms and formulae.

danieldietsch commented on trunk/source/Library-CHC/src/de/uni_freiburg/informatik/ultimate/lib/chc/eldarica/EldaricaChcScript.java:
```diff
@@ -0,0 +1,370 @@
+				if (actualTimeout.isDefined() && System.currentTimeMillis() - startTime > actualTimeout.get()) {
+					// Nasty hack to trick java into throwing TimeoutException, a checked exception.
+					// (This is necessary, because scala does not declare checked exceptions.)
+					throwUnchecked(new lazabs.Main.TimeoutException$());
```
Uhh, the pain.

maul-esel replied:
:smiling_imp:

danieldietsch commented on trunk/source/Library-CHC/src/de/uni_freiburg/informatik/ultimate/lib/chc/results/ChcSatResult.java:
```diff
@@ -0,0 +1,82 @@
+ * @author Dominik Klumpp (klumpp@informatik.uni-freiburg.de)
+ *
+ */
+public class ChcSatResult implements IResult {
```
It might be ok to have a ChcResult and just subclass the three from that ...

danieldietsch commented on trunk/source/Library-UltimateTest/src/de/uni_freiburg/informatik/ultimate/test/decider/ChcTestResultDecider.java:
```diff
@@ -31,53 +31,53 @@
-				mCategory = "precise match of results (and results are not both UNKOWN..)";
-				mMessage = "both results are" + expectedResult;
+				mCategory = "precise match of results (and results are not both UNKOWN).";
+				mMessage = "both results are " + expectedResult;
```
👍

danieldietsch commented on trunk/source/TreeAutomizer/src/de/uni_freiburg/informatik/ultimate/plugins/generator/treeautomizer/TreeAutomizerChcScript.java:
```diff
@@ -0,0 +1,175 @@
+			final List<HornClause> system) {
+		reset();
+
+		// TODO missing parameter: category info
```
?

maul-esel replied:
The last argument of the `HornAnnot` constructor in the next line.

danieldietsch replied:
I just dont know if this is an issue and what the parameter does. But if it is ok to leave it like that, just leave it.

maul-esel replied:
Resolved in ef5fd2516e

danieldietsch commented on trunk/source/TreeAutomizer/src/de/uni_freiburg/informatik/ultimate/plugins/generator/treeautomizer/TreeAutomizerChcScript.java:
```diff
@@ -0,0 +1,175 @@
+
+	@Override
+	public boolean supportsModelProduction() {
+		// TODO TreeAutomizer seems to produce some kind model, but just logs it (instead of returning it)
```
@Heizmann do you know anything about that?

danieldietsch reviewed (approved):
But please rebase instead of merge ;p

maul-esel committed:
better exceptions (@danieldietsch)

maul-esel committed:
small code simplification (@danieldietsch)

maul-esel committed:
improved documentation (@danieldietsch)

maul-esel committed:
reorganize regression tests

ChcSolver only runs on examples simple enough for all solvers
to finish in time, such that only the integration is tested.
TreeAutomizer still runs on all examples.

maul-esel committed:
improve documentation

maul-esel commented:
@danieldietsch: Sorry:

![Screenshot from 2023-05-19 09-15-44](https://github.com/ultimate-pa/ultimate/assets/800556/acf0e1d4-9ac5-4d18-974d-f6d42d00eac7)


I still don't understand your issue with merge commits, though.

danieldietsch commented:
@maul-esel ... we can discuss this, but if you like you can read, e.g., https://www.atlassian.com/git/tutorials/merging-vs-rebasing
